### PR TITLE
Add get_stdout and get_resources method for the task class

### PR DESCRIPTION
**Pull request by david451lima**
Added get_stdout() and get_resources() methods to the task class.

For a task that is running, either function returns the last n_lines (passed as an argument to both functions) of the stdout_live.txt and resources_usage.txt files, respectively.
The output is presented as a list of lines.

The functions makes a call to the get_file_tail endpoint, specifying the file corresponding to the information we want to query.

Test example:

```
pdb_file = "/Users/davidlima/Desktop/MD-datasets/alanine-dataset/alanine.pdb"

scenario = inductiva.molecules.ProteinSolvation(pdb_file)


task = scenario.simulate(simulation_time_ns=11.01,
                             n_steps_min = 100, run_async = True)#,
                            # machine_group = machine)

print(task.get_resources(n_lines = 5))
```
returns:
```
INFO:absl:Task ID: 1697474928594377086
INFO:absl:Creating input archive...
INFO:absl:Input archive created.
INFO:absl:Input archive size: 3.20 KiB
INFO:absl:Uploading input...
INFO:absl:Input archive uploaded.
INFO:absl:Task request submitted.
('2023-10-16T16:49:14.462134334Z, 4.040772184821457, 82.11372151898733 ', '2023-10-16T16:49:15.474630085Z, 4.838320451338234, 78.1439012345679 ', '2023-10-16T16:49:16.479307585Z, 5.234860138575377, 
67.83508771929824 ', '2023-10-16T16:49:17.483268419Z, 3.382951276166604, 74.70852867830425 ', '2023-10-16T16:49:18.489597128Z, 4.417847364495745, 85.22639386189257 ')
```
and

```
pdb_file = "/Users/davidlima/Desktop/MD-datasets/alanine-dataset/alanine.pdb"

scenario = inductiva.molecules.ProteinSolvation(pdb_file)


task = scenario.simulate(simulation_time_ns=11.01,
                             n_steps_min = 100, run_async = True)#,
                            # machine_group = machine)
                            
print(task.get_stdout(n_lines = 5))
```
returns:

```
INFO:absl:Task ID: 1697481254418783514
INFO:absl:Creating input archive...
INFO:absl:Input archive created.
INFO:absl:Input archive size: 3.20 KiB
INFO:absl:Uploading input...
INFO:absl:Input archive uploaded.
INFO:absl:Task request submitted.
('', '', ' -------', '# COMMAND: gmx grompp -f energy_minimization.mdp -c protein_solv_ions.gro -p topol.top -o em.tpr -maxwarn 0 ', '')
```

EDIT: I seem to be getting errors with the pytest in the tail function definition in the client (inductiva/client/paths/tail/get.py). I don't understand quite what is the problem, since the code runs quite well. Can someone check this.

### CODE REVIEW

File: inductiva/tasks/task.py [hunk 1]
```diff
@@ -49,6 +50,7 @@ def __init__(self, task_id: str):
         """Initialize the instance from a task ID."""
         self.id = task_id
         self._api = tasks_api.TasksApi(api.get_client())
+        self._api_default = default_api.DefaultApi(api.get_client())
```
Comment (IvanPombo) [inline inductiva/tasks/task.py:53]: What is this for and what will be the difference for the _api?

File: inductiva/tasks/task.py [hunk 2]
```diff
@@ -408,3 +410,58 @@ def get_machine_type(self) -> Optional[str]:
...
+
+        Args:
+            n_lines: Number of lines to return from the end of the file.
+        Returns:
+            A list of strings, each string being a line from the stdout."""
+
+        #This path needs to be updated with the self.parent_dir
+        # when that feature is added to tasks
+        path = os.path.join(self.id, "stdout_live.txt")
+        api_response = self._api_default.get_file_tail(
```
[the remark this answers is not in the document]
Comment (david451lima) [reply inductiva/tasks/task.py:430]: The endpoint can be used for anyfile. This function only specifies the files of stdout and resources

File: inductiva/tasks/task.py [hunk 3]
```diff
@@ -408,3 +410,58 @@ def get_machine_type(self) -> Optional[str]:
         machine_type = machine_info["vm_type"].split("/")[-1]
 
         return machine_type
+
+    def get_stdout(self, n_lines: int = 10):
+        """Returns tail of stdout.txt file for current task
+
+        Calls the get_file_tail function, specifying the path of the
+        stdout_live.txt file, that contains the live version of the
```
Comment (cristianacarpinteiro) [inline inductiva/tasks/task.py:418]: maybe say the most recent version instead of live version

File: inductiva/tasks/task.py [hunk 4]
```diff
@@ -408,3 +410,58 @@ def get_machine_type(self) -> Optional[str]:
...
+        api_response = self._api_default.get_file_tail(
+            query_params={
+                "path": path,
+                "n_lines": n_lines,
+            },
+            stream=False,
+            skip_deserialization=False,
+        )
+
+        return api_response.body
```
Comment (cristianacarpinteiro) [inline inductiva/tasks/task.py:439]: Should we print this in a pretty way to be more readable by the user?

File: inductiva/tasks/task.py [hunk 5]
```diff
@@ -408,3 +410,58 @@ def get_machine_type(self) -> Optional[str]:
...
+
+    def get_resources(self, n_lines: int = 10):
```
Comment (cristianacarpinteiro) [inline inductiva/tasks/task.py:441]: maybe call it get_resource_usage?

File: inductiva/tasks/task.py [hunk 6]
```diff
@@ -408,3 +410,58 @@ def get_machine_type(self) -> Optional[str]:
...
+        lines.
+
+        Args:
+            n_lines: Number of lines to return from the end of the file.
+        Returns:
+            List of lines from the end of the resource_usage file."""
+
+        #This path needs to be updated with the self.parent_dir
+        # when that feature is added to tasks
+        path = os.path.join(self.id, "resource_usage.txt")
```
Comment (cristianacarpinteiro) [inline inductiva/tasks/task.py:457]: you should update this, I merged my PR just now.

File: inductiva/tasks/task.py [hunk 7]
```diff
@@ -408,3 +410,58 @@ def get_machine_type(self) -> Optional[str]:
...
+        of lines.
+
+        Args:
+            n_lines: Number of lines to return from the end of the file.
+        Returns:
+            A list of strings, each string being a line from the stdout."""
+
+        #This path needs to be updated with the self.parent_dir
+        # when that feature is added to tasks
+        path = os.path.join(self.id, "stdout_live.txt")
```
[the remark this answers is not in the document]
Comment (david451lima) [reply inductiva/tasks/task.py]: The code is updated with the new feature that Cristiana implemented.
I understand your concerns, but the implementation you suggest requires the stdout file to be tracked at all times, even in non related functions like the ones that change the name of the files.
Since this function is to be mainly used as the simulation runs, to check its status, I don't think that we need to adress that issue. Wdyt?

[the remark this answers is not in the document]
Comment (david451lima) [reply inductiva/tasks/task.py]: Tracked as in you need to register somewhere what is the "stdout" file that you want to display when you call get_stdout in the client. That would require checks to the function change_filename, that checks if we are changing the name / moving the stdout file

[the remark this answers is not in the document]
Comment (david451lima) [reply inductiva/tasks/task.py]: That approach (get the stdout in the backend by specifying the path there) was the initial draft of this method, but @hpenedones pointed out that, for the sake of modularity, we should have the generic tail endpoint and pass the specific path of stdout.txt later.

If I pass it to the backend the changes in future modifications of the way we read stdout will be the same regardless of where I have the function (change the name for example).

[the remark this answers is not in the document]
Comment (hpenedones) [reply inductiva/tasks/task.py]: I think what @luispcunha is suggesting, which makes sense: is that the client should not need to be aware of those filenames that are stored in the backend (which could potentially be changed in the future).

You can have the get stdout of a task implemented in the backend, while still respecting what I suggested: implement first a function in the backend that can head or tail the content of any file (done!), and then have an backend endpoint for a task stdout calls that function. In the client package you can just call that backend endpoint directly. 

In other words: I basically agree with everything @luispcunha said in its first comment in this discussion (what you did is "almost there", you just need to move this to the backend and in the client call it with one line)

[the remark this answers is not in the document]
Comment (david451lima) [reply inductiva/tasks/task.py]: > is that the client should not need to be aware of those filenames that are stored in the backend
But the client is not aware of that, since it only calls get_stdout(), and never specifies the path.

What I don't understand is the need of having three function (standard tail -> tail for the stdout -> get_stdout in the client) instead of the cleaner two (standard tail -> get_stdout in the client)

[the remark this answers is not in the document]
Comment (david451lima) [reply inductiva/tasks/task.py]: Ok I understand the logic. I though that the way it is implemented is cleaner, since it has only two dependencies (tail endpoint -> get_stdout method) instead of three (tail endpoint -> tail endpoint for stdout -> get_stdout method), but the points raised by Luis also make sense

[the remark this answers is not in the document]
Comment (david451lima) [reply inductiva/tasks/task.py]: Makes perfect sense! Thanks
